metrics-graphite/src/main/java/com/codahale/metrics/graphite/PickledGraphiteUDP.java

```diff
@@ -0,0 +1,203 @@
+     *            the port of the Carbon server
+     */
+    public PickledGraphiteUDP(String hostname, int port) {
+        this(hostname, port, DEFAULT_BATCH_SIZE);
+    }
+
+    /**
+     * Creates a new client which connects to the given address and socket factory.
+     *
+     * @param hostname
+     *            the hostname of the Carbon server
+     * @param port
+     *            the port of the Carbon server
+     * @param batchSize
+     *            how many metrics are bundled into a single pickle request to graphite
+     */
+    public PickledGraphiteUDP(String hostname, int port, int batchSize) {
+        this(hostname, port, UTF_8, batchSize);
+    }
+
+    /**
+     * Creates a new client which connects to the given address and socket factory using the given character set.
+     *
+     * @param hostname
+     *            the hostname of the Carbon server
+     * @param port
+     *            the port of the Carbon server
+     * @param charset
+     *            the character set used by the server
+     * @param batchSize
+     *            how many metrics are bundled into a single pickle request to graphite
+     */
+    public PickledGraphiteUDP(String hostname, int port, Charset charset, int batchSize) {
+        this(new InetSocketAddress(hostname, port), charset, batchSize);
+    }
+
+    @Override
+    public void connect() throws IllegalStateException, IOException {
+        if (isConnected()) {
+            throw new IllegalStateException("Already connected");
+        }
+        InetSocketAddress address = this.address;
+        if (address.getAddress() == null) {
+            throw new UnknownHostException(address.getHostName());
+        }
+
+        datagramChannel = DatagramChannel.open();
+        datagramChannel.connect(address);
+    }
+
+    @Override
+    public boolean isConnected() {
+        return datagramChannel != null && !datagramChannel.socket().isClosed();
+    }
+
+    /**
+     * Convert the metric to a python tuple of the form:
+     * <p/>
+     * (timestamp, (name, value))
+     * <p/>
+     * And add it to the list of metrics. If we reach the batch size, write them out.
+     *
+     * @param name
+     *            the name of the metric
+     * @param value
+     *            the value of the metric
+     * @param timestamp
+     *            the timestamp of the metric
+     * @throws IOException
+     *             if there was an error sending the metric
+     */
+    @Override
+    public void send(String name, String value, long timestamp) throws IOException {
+        metrics.add(new GraphitePickler.MetricTuple(name, timestamp, value));
+
+        if (metrics.size() >= batchSize) {
+            writeMetrics();
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        // No such concept in UDP
+    }
+
+    @Override
+    public void close() throws IOException {
+        datagramChannel.socket().close();
+        datagramChannel = null;
```

Comment:
I think it's better to dereference the channel in a finally block in case the socket failed to close.